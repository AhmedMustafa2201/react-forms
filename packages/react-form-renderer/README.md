# React Form Renderer

Rendering forms from data structures

## Documentation: http://data-driven-forms.org/
