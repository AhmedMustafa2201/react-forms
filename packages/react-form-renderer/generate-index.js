const glob = require('glob');
const path = require('path');
const fse = require('fs-extra');

const src = path.resolve(__dirname, './src');

function camelize(str) {
  const arr = str.split('-');
  const capital = arr.map((item, index) => (index ? item.charAt(0).toUpperCase() + item.slice(1).toLowerCase() : item.toLowerCase()));
  return capital.join('');
}

function generateIndex() {
  const directories = glob
    .sync(`${src}/*/`)
    .filter((name) => !name.includes('/tests/') && !name.includes('/common/'))
    .map((path) =>
      path
        .replace(/\/$/, '')
        .split('/')
        .pop()
    );
  const content = `${directories.map(
    (dir) => `export { default as ${camelize(dir)} } from './${dir}';
`
  )}`.replace(/,/g, '');

  return fse.writeFileSync(`${src}/index.js`, content);
}

function run() {
  try {
    generateIndex();
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}

run();
