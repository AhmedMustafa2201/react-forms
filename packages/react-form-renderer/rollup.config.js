import nodeResolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import replace from 'rollup-plugin-replace';
import nodeGlobals from 'rollup-plugin-node-globals';
import { terser } from 'rollup-plugin-terser';
import { sizeSnapshot } from 'rollup-plugin-size-snapshot';

const globals = {
  react: 'React',
  'react-dom': 'ReactDOM',
};

const babelOptions = {
  exclude: /node_modules/,
  runtimeHelpers: true,
  configFile: '../../babel.config.js',
};

const commonjsOptions = {
  ignoreGlobal: true,
  include: /node_modules/,
  namedExports: {
    './src/form-renderer/helpers': [ 'composeValidators' ],
  },
};

function onwarn(warning) {
  throw Error(warning.message);
}

export default [{
  onwarn,
  input: './src/index.js',
  output: {
    file: './dist/index.js',
    format: 'umd',
    name: '@data-driven-forms/react-form-renderer',
    exports: 'named',
    globals,
  },
  external: Object.keys(globals),
  plugins: [
    nodeResolve(),
    babel(babelOptions),
    commonjs(commonjsOptions),
    nodeGlobals(), // Wait for https://github.com/cssinjs/jss/pull/893
    replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
    sizeSnapshot({ snapshotPath: 'size-snapshot.json' }),
    terser({
      keep_classnames: true,
      keep_fnames: true,
    }),
  ],
}];
