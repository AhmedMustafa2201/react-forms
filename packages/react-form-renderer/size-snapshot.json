{
  "dist/index.js": {
    "bundled": 246500,
    "minified": 87931,
    "gzipped": 26546
  }
}
