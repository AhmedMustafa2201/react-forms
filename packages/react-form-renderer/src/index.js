export { default as commonTypes } from './common-types';
export { default as componentTypes } from './component-types';
export { default as composeValidators } from './compose-validators';
export { default as condition } from './condition';
export { default as convertType } from './convert-type';
export { default as dataTypes } from './data-types';
export { default as defaultSchemaValidator } from './default-schema-validator';
export { default as fieldArray } from './field-array';
export { default as fieldProvider } from './field-provider';
export { default as formRenderer } from './form-renderer';
export { default as formSpy } from './form-spy';
export { default as form } from './form';
export { default as parseCondition } from './parse-condition';
export { default as rendererContext } from './renderer-context';
export { default as schemaErrors } from './schema-errors';
export { default as useFieldApi } from './use-field-api';
export { default as useFormApi } from './use-form-api';
export { default as validatorMapper } from './validator-mapper';
export { default as validatorTypes } from './validator-types';
export { default as validators } from './validators';
export { default as wizardContext } from './wizard-context';
