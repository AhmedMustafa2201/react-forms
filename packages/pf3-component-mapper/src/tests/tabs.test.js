import React from 'react';
import { mount } from 'enzyme';
import { componentTypes } from '@data-driven-forms/react-form-renderer';
import { NavItem } from 'patternfly-react';
import toJson from 'enzyme-to-json';
import FormTabs from '../components/tabs';

describe('<FormTabs />', () => {
  let initialProps;
  beforeEach(() => {
    initialProps = {
      fields: [{
        component: componentTypes.TABS,
        title: 'Tab 1',
        name: 'tab1',
        fields: [{
          name: 'foo',
          component: 'foo',
        }],
      }, {
        component: componentTypes.TABS,
        title: 'Tab 2',
        name: 'tab2',
        fields: [],
      }],
      formOptions: {
        renderForm: ({ name, component }) => <div key={ name }>{ component }</div>,
        getState: () => ({
          errors: {},
        }),
      },
    };
  });

  it('should render form tabs', () => {
    const wrapper = mount(<FormTabs { ...initialProps } />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render form tabs with error state', () => {
    const validationSchema = {
      ...initialProps,
      fields: [{
        component: componentTypes.TABS,
        title: 'Tab 1',
        name: 'tab1',
        validateFields: [ 'foo', 'nested.field' ],
        fields: [{
          name: 'foo',
          component: 'foo',
        }, {
          name: 'nested.field',
          component: 'foo',
        }],
      }],
      formOptions: {
        renderForm: ({ name, component }) => <div key={ name }>{ component }</div>,
        getState: () => ({
          errors: { foo: true, nested: { field: true }},
        }),
      },
    };
    const wrapper = mount(<FormTabs { ...validationSchema } />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should switch form tabs on click', () => {
    const wrapper = mount(<FormTabs { ...initialProps } />);
    expect(wrapper.find(NavItem).first().instance().props.active).toEqual(true);
    expect(wrapper.find(NavItem).last().instance().props.active).toEqual(false);

    wrapper.find('a').last().simulate('click');
    expect(wrapper.find(NavItem).first().instance().props.active).toEqual(false);
    expect(wrapper.find(NavItem).last().instance().props.active).toEqual(true);
  });
});
