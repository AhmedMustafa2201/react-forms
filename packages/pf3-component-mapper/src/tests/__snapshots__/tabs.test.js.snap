// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormTabs /> should render form tabs 1`] = `
<FormTabs
  fields={
    Array [
      Object {
        "component": "tabs",
        "fields": Array [
          Object {
            "component": "foo",
            "name": "foo",
          },
        ],
        "name": "tab1",
        "title": "Tab 1",
      },
      Object {
        "component": "tabs",
        "fields": Array [],
        "name": "tab2",
        "title": "Tab 2",
      },
    ]
  }
  formOptions={
    Object {
      "getState": [Function],
      "renderForm": [Function],
    }
  }
>
  <ForwardRef
    defaultActiveKey={0}
    id="basic-tabs-pf"
  >
    <Uncontrolled(TabContainer)
      defaultActiveKey={0}
      id="basic-tabs-pf"
      innerRef={null}
    >
      <TabContainer
        activeKey={0}
        id="basic-tabs-pf"
        onSelect={[Function]}
      >
        <div
          id="basic-tabs-pf"
        >
          <Nav
            bsClass="nav nav-tabs"
            justified={false}
            pullLeft={false}
            pullRight={false}
            stacked={false}
          >
            <ul
              className="nav nav-tabs"
              role="tablist"
            >
              <NavItem
                active={true}
                activeKey={0}
                aria-controls="basic-tabs-pf-pane-0"
                disabled={false}
                eventKey={0}
                id="basic-tabs-pf-tab-0"
                key=".$tab1"
                onKeyDown={[Function]}
                onSelect={[Function]}
                role="tab"
              >
                <li
                  className="active"
                  role="presentation"
                >
                  <SafeAnchor
                    aria-controls="basic-tabs-pf-pane-0"
                    aria-selected={true}
                    componentClass="a"
                    disabled={false}
                    id="basic-tabs-pf-tab-0"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="tab"
                  >
                    <a
                      aria-controls="basic-tabs-pf-pane-0"
                      aria-selected={true}
                      href="#"
                      id="basic-tabs-pf-tab-0"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="tab"
                    >
                      Tab 1
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey={0}
                aria-controls="basic-tabs-pf-pane-1"
                disabled={false}
                eventKey={1}
                id="basic-tabs-pf-tab-1"
                key=".$tab2"
                onKeyDown={[Function]}
                onSelect={[Function]}
                role="tab"
                tabIndex={-1}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    aria-controls="basic-tabs-pf-pane-1"
                    aria-selected={false}
                    componentClass="a"
                    disabled={false}
                    id="basic-tabs-pf-tab-1"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="tab"
                    tabIndex={-1}
                  >
                    <a
                      aria-controls="basic-tabs-pf-pane-1"
                      aria-selected={false}
                      href="#"
                      id="basic-tabs-pf-tab-1"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="tab"
                      tabIndex={-1}
                    >
                      Tab 2
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
            </ul>
          </Nav>
          <TabContent
            animation={true}
            bsClass="tab"
            componentClass="div"
            mountOnEnter={false}
            unmountOnExit={false}
          >
            <div
              className="tab-content"
            >
              <TabPane
                bsClass="tab-pane"
                eventKey={0}
                key="tab1"
              >
                <Fade
                  appear={false}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onExited={[Function]}
                  timeout={300}
                  unmountOnExit={false}
                >
                  <Transition
                    appear={false}
                    enter={true}
                    exit={true}
                    in={true}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    timeout={300}
                    unmountOnExit={false}
                  >
                    <div
                      aria-hidden={false}
                      aria-labelledby="basic-tabs-pf-tab-0"
                      className="fade tab-pane active in"
                      id="basic-tabs-pf-pane-0"
                      role="tabpanel"
                    >
                      <div />
                    </div>
                  </Transition>
                </Fade>
              </TabPane>
              <TabPane
                bsClass="tab-pane"
                eventKey={1}
                key="tab2"
              >
                <Fade
                  appear={false}
                  in={false}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onExited={[Function]}
                  timeout={300}
                  unmountOnExit={false}
                >
                  <Transition
                    appear={false}
                    enter={true}
                    exit={true}
                    in={false}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    timeout={300}
                    unmountOnExit={false}
                  >
                    <div
                      aria-hidden={true}
                      aria-labelledby="basic-tabs-pf-tab-1"
                      className="fade tab-pane"
                      id="basic-tabs-pf-pane-1"
                      role="tabpanel"
                    >
                      <div />
                    </div>
                  </Transition>
                </Fade>
              </TabPane>
            </div>
          </TabContent>
        </div>
      </TabContainer>
    </Uncontrolled(TabContainer)>
  </ForwardRef>
</FormTabs>
`;

exports[`<FormTabs /> should render form tabs with error state 1`] = `
<FormTabs
  fields={
    Array [
      Object {
        "component": "tabs",
        "fields": Array [
          Object {
            "component": "foo",
            "name": "foo",
          },
          Object {
            "component": "foo",
            "name": "nested.field",
          },
        ],
        "name": "tab1",
        "title": "Tab 1",
        "validateFields": Array [
          "foo",
          "nested.field",
        ],
      },
    ]
  }
  formOptions={
    Object {
      "getState": [Function],
      "renderForm": [Function],
    }
  }
>
  <ForwardRef
    defaultActiveKey={0}
    id="basic-tabs-pf"
  >
    <Uncontrolled(TabContainer)
      defaultActiveKey={0}
      id="basic-tabs-pf"
      innerRef={null}
    >
      <TabContainer
        activeKey={0}
        id="basic-tabs-pf"
        onSelect={[Function]}
      >
        <div
          id="basic-tabs-pf"
        >
          <Nav
            bsClass="nav nav-tabs"
            justified={false}
            pullLeft={false}
            pullRight={false}
            stacked={false}
          >
            <ul
              className="nav nav-tabs"
              role="tablist"
            >
              <NavItem
                active={true}
                activeKey={0}
                aria-controls="basic-tabs-pf-pane-0"
                disabled={false}
                eventKey={0}
                id="basic-tabs-pf-tab-0"
                key=".$tab1"
                onKeyDown={[Function]}
                onSelect={[Function]}
                role="tab"
              >
                <li
                  className="active"
                  role="presentation"
                >
                  <SafeAnchor
                    aria-controls="basic-tabs-pf-pane-0"
                    aria-selected={true}
                    componentClass="a"
                    disabled={false}
                    id="basic-tabs-pf-tab-0"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="tab"
                  >
                    <a
                      aria-controls="basic-tabs-pf-pane-0"
                      aria-selected={true}
                      href="#"
                      id="basic-tabs-pf-tab-0"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="tab"
                    >
                      <Icon
                        name="exclamation-circle"
                        style={
                          Object {
                            "color": "#CC0000",
                            "marginRight": 8,
                          }
                        }
                        type="fa"
                      >
                        <FontAwesome
                          name="exclamation-circle"
                          style={
                            Object {
                              "color": "#CC0000",
                              "marginRight": 8,
                            }
                          }
                        >
                          <span
                            aria-hidden={true}
                            className="fa fa-exclamation-circle"
                            style={
                              Object {
                                "color": "#CC0000",
                                "marginRight": 8,
                              }
                            }
                          />
                        </FontAwesome>
                      </Icon>
                      Tab 1
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
            </ul>
          </Nav>
          <TabContent
            animation={true}
            bsClass="tab"
            componentClass="div"
            mountOnEnter={false}
            unmountOnExit={false}
          >
            <div
              className="tab-content"
            >
              <TabPane
                bsClass="tab-pane"
                eventKey={0}
                key="tab1"
              >
                <Fade
                  appear={false}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onExited={[Function]}
                  timeout={300}
                  unmountOnExit={false}
                >
                  <Transition
                    appear={false}
                    enter={true}
                    exit={true}
                    in={true}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    timeout={300}
                    unmountOnExit={false}
                  >
                    <div
                      aria-hidden={false}
                      aria-labelledby="basic-tabs-pf-tab-0"
                      className="fade tab-pane active in"
                      id="basic-tabs-pf-pane-0"
                      role="tabpanel"
                    >
                      <div />
                    </div>
                  </Transition>
                </Fade>
              </TabPane>
            </div>
          </TabContent>
        </div>
      </TabContainer>
    </Uncontrolled(TabContainer)>
  </ForwardRef>
</FormTabs>
`;
